go/libkb/pvl_interp.go

```diff
@@ -0,0 +1,825 @@
+		return NewProofError(keybase1.ProofStatus_DNS_ERROR,
+			"DNS failure for %s: %s", domain, err)
+	}
+
+	for _, record := range txts {
+		g.GetLog().Debug("For %s, got TXT record: %s", domain, record)
+
+		// Try all scripts.
+		for _, script := range scripts {
+			state := startstate
+			state.ActiveString = record
+			err = pvlRunScript(g, script, state)
+			if err == nil {
+				return nil
+			}
+		}
+	}
+
+	return NewProofError(keybase1.ProofStatus_NOT_FOUND,
+		"Checked %d TXT entries of %s, but didn't find signature",
+		len(txts), domain)
+}
+
+func pvlRunScript(g GlobalContextLite, script *jsonw.Wrapper, startstate PvlScriptState) ProofError {
+	var state = startstate
+	scriptlen, err := script.Len()
+	if err != nil {
+		return NewProofError(keybase1.ProofStatus_INVALID_PVL,
+			"Could not get length of script: %v", err)
+	}
+	if scriptlen < 1 {
+		return NewProofError(keybase1.ProofStatus_INVALID_PVL,
+			"Empty scripts are not allowed.")
+	}
+	for i := 0; i < scriptlen; i++ {
+		ins := script.AtIndex(i)
+
+		// Sanity check.
+		if int(state.PC) != i {
+			return NewProofError(keybase1.ProofStatus_INVALID_PVL,
+				fmt.Sprintf("Execution failure, PC mismatch %v %v", state.PC, i))
+		}
+
+		newstate, perr := pvlStepInstruction(g, ins, state)
+		state = newstate
+		if perr != nil {
+			// Add additional info to INVALID_PVL errors.
+			if perr.GetProofStatus() == keybase1.ProofStatus_INVALID_PVL {
+				perr = NewProofError(keybase1.ProofStatus_INVALID_PVL,
+					fmt.Sprintf("Invalid PVL (%v): %v", state.PC, perr.GetDesc()))
+			}
+			return perr
+		}
+		state.PC++
+	}
+
+	// Script executed successfully with no errors.
+	return nil
+}
+
+func pvlStepInstruction(g GlobalContextLite, ins *jsonw.Wrapper, state PvlScriptState) (PvlScriptState, ProofError) {
+	for name, step := range PvlSteps {
+		if pvlJSONHasKey(ins, name) {
+			return step(g, ins, state)
+		}
+	}
+
+	return state, NewProofError(keybase1.ProofStatus_INVALID_PVL,
+		"Unsupported PVL instruction %d", state.PC)
+}
+
+func pvlStepAssertRegexMatch(g GlobalContextLite, ins *jsonw.Wrapper, state PvlScriptState) (PvlScriptState, ProofError) {
+	template, err := ins.AtKey(PvlAssertRegexMatch).GetString()
+	if err != nil {
+		return state, NewProofError(keybase1.ProofStatus_INVALID_PVL,
+			"Could not get template %v", ins)
+	}
+	re, perr := pvlInterpretRegex(template, state.Vars)
+	if perr != nil {
+		return state, perr
+	}
+	if !re.MatchString(state.ActiveString) {
+		g.GetLog().Debug("PVL regex did not match: %v %v", re, state.ActiveString)
+		return state, NewProofError(keybase1.ProofStatus_CONTENT_FAILURE,
+			"Regex did not match %v", re)
+	}
+
+	return state, nil
+}
+
+func pvlStepAssertFindBase64(g GlobalContextLite, ins *jsonw.Wrapper, state PvlScriptState) (PvlScriptState, ProofError) {
+	target, err := ins.AtKey(PvlAssertFindBase64).GetString()
+	if err != nil {
+		return state, NewProofError(keybase1.ProofStatus_INVALID_PVL,
+			"Could not assert target %v", ins)
```

Comment:
Let's include the why part here. `...because only 'sig' is supported.` or something.